pod-overhead: updates for beta (#19059)
kubernetes/website: content/en/docs/concepts/configuration/pod-overhead.md

        cpu: "250m"
```

Workloads which are created which specify the `kata-fc` RuntimeClass handler will take the memory and
cpu overheads into account for resource quota calculations, node scheduling, as well as Pod cgroup sizing.

Consider running the given example workload, test-pod:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  runtimeClassName: kata-fc
  containers:
  - name: busybox-ctr
    image: busybox
    stdin: true
    tty: true
    resources:
      limits:
        cpu: 500m
        memory: 100Mi
  - name: nginx-ctr
    image: nginx
    resources:
      limits:
        cpu: 1500m
        memory: 100Mi
```

At admission time the RuntimeClass [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/)
updates the workload's PodSpec to include the `overhead` as described in the RuntimeClass. If the PodSpec already has this field defined,
the Pod will be rejected. In the given example, since only the RuntimeClass name is specified, the admission controller mutates the Pod
to include an `overhead`.

After the RuntimeClass admission controller, you can check the updated PodSpec:

```bash

> kubectl get pod test-pod -o jsonpath='{.spec.overhead}'

Reviewer: Is pod overhead included in the output of `kubectl describe`? If so, that might be a more user-friendly way of checking

Author: It is not part of kubectl describe.

```

The output is:
```
map[cpu:250m memory:120Mi]
```

If a ResourceQuota is defined, the sum of container requests as well as the
`overhead` field are counted.

When the kube-scheduler is deciding which node should run a new Pod, the scheduler considers that Pod's
`overhead` as well as the sum of container requests for that Pod. For this example, the scheduler adds the
requests and the overhead, then looks for a node that has 2.25 CPU and 320 MiB of memory available.

Once a Pod is scheduled to a node, the kubelet on that node creates a new {{< glossary_tooltip text="cgroup" term_id="cgroup" >}}
for the Pod. It is within this pod that the underlying container runtime will create containers.

If the resource has a limit defined for each container (Guaranteed QoS or Bustrable QoS with limits defined),
the kubelet will set an upper limit for the pod cgroup associated with that resource (cpu.cfs_quota_us for CPU
and memory.limit_in_bytes memory). This upper limit is based on the sum of the container limits plus the `overhead`
defined in the PodSpec.

> For CPU, if the Pod is Quaranteed or Bustrable QoS, kubelet will set cpu.shares based on the sum of container

Reviewer: ```suggestion
For CPU, if the Pod is Guaranteed or Bustrable QoS, kubelet will set cpu.shares based on the sum of container
```

Reviewer 2: Actually there's 2 typos and I can see some other things the style guide has an opinion on:
```suggestion
For CPU, if the Pod is Guaranteed or Burstable QoS, the kubelet sets `cpu.shares` based on the sum of container
```

Author: Quaranteed.... too much quarantine discussion at the water cooler i think...

requests plus the `overhead` defined in the PodSpec.

Looking at our example, verify the container requests for the workload:
```bash
kubectl get pod test-pod -o jsonpath='{.spec.containers[*].resources.limits}'
```

The total container requests are 2000m CPU and 200MiB of memory:
```
map[cpu: 500m memory:100Mi] map[cpu:1500m memory:100Mi]
```

Check this against what is observed by the node:
```bash
kubectl describe node | grep test-pod -B2
```

The output shows 2250m CPU and 320MiB of memory are requested, which includes PodOverhead:
```
  Namespace                   Name                CPU Requests  CPU Limits   Memory Requests  Memory Limits  AGE
  ---------                   ----                ------------  ----------   ---------------  -------------  ---
  default                     test-pod            2250m (56%)   2250m (56%)  320Mi (1%)       320Mi (1%)     36m
```

> Check the Pod's memory cgroups on the node where the workload is running. First, on the particular node, determine

Reviewer: I feel like this should go in a new header. This is a more advanced topic, and most users won't need to go through the effort to verify this (if the feature is working correctly). Maybe `## Verify Pod cgroup limits` ?

the Pod identifier:

```bash
# Run this on the node where the Pod is scheduled

> sudo crictl pods | grep test-pod

Reviewer: how about this instead:
```suggestion
POD_ID=$(sudo crictl pods --name test-pod -q)
```

Then below, you can just use `$POD_ID`

Reviewer 2: nit (modelling safer coding style):
```suggestion
POD_ID="$(sudo crictl pods --name test-pod -q)"
```
?

> sudo crictl pods | grep test-pod

Reviewer: I feel like there should be some note introducing crictl. Maybe with a link to where to get the tool. A lot of clusters probably still rely on docker.

Author: sure.  I suppose of they are using PodOverhead, they are probably using an alternative runtime handler and would have had to have moved on from Docker anyway...

```

> The value observed is `7ccf55aee35dd`:

Reviewer: ```suggestion
The pod's POD ID `7ccf55aee35dd`:
```

Author: Instead I'll just remove this part of the example, and use POD_ID in the inspection below.

{{% capture body %}}

In Kubernetes, the Pod's overhead is set at
[admission](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks)
time according to the overhead associated with the Pod's
[RuntimeClass](/docs/concepts/containers/runtime-class/).

When Pod Overhead is enabled, the overhead is considered in addition to the sum of container
resource requests when scheduling a Pod. Similarly, Kubelet will include the Pod overhead when sizing
the Pod cgroup, and when carrying out Pod eviction ranking.

## Enabling Pod Overhead {#set-up}

You need to make sure that the `PodOverhead`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled (it is on by default as of 1.18)

> across your cluster, and a `RuntimeClass` is utilized which defines the `Overhead` field.

Reviewer 2: ```suggestion
across your cluster, and a `RuntimeClass` is utilized which defines the `overhead` field.
```
(matches the example manifest)

Author: done.

## Pod Overhead

In Kubernetes, the Pod's overhead is set at
[admission](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks)
time according to the overhead associated with the Pod's
[RuntimeClass](/docs/concepts/containers/runtime-class/).

When Pod Overhead is enabled, the overhead is considered in addition to the sum of container
resource requests when scheduling a Pod. Similarly, Kubelet will include the Pod overhead when sizing
the Pod cgroup, and when carrying out Pod eviction ranking.

### Set Up

You need to make sure that the `PodOverhead`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled (it is on by default as of 1.18)
across your cluster, and a `RuntimeClass` is utilized which defines the `Overhead` field.

### Usage example

To use the `PodOverhead` feature, a `RuntimeClass` must be utilized which defines the `Overhead` field. As an example,

> the following `RuntimeClass` definition could be utilized for Kata Containers:

Reviewer 2: nit / rewording to generalize:
```suggestion
To use the PodOverhead feature, you need a RuntimeClass that defines the `overhead` field. As
an example, you could use the following RuntimeClass definition with a virtualizing container runtime
that uses around 120MiB per Pod for the virtual machine and the guest OS:
```

```yaml
---
kind: RuntimeClass
apiVersion: node.k8s.io/v1beta1
metadata:
    name: kata-fc
handler: kata-fc
overhead:
    podFixed:
        memory: "120Mi"
        cpu: "250m"

> ```

Reviewer 2: Consider making this a downloadable example; see https://kubernetes.io/docs/contribute/style/write-new-topic/#embedding-code-in-your-topic for details (use `{{< codenew >}}`).

Workloads which are created which specify the `kata-fc` RuntimeClass handler will take the memory and
cpu overheads into account for resource quota calculations, node scheduling, as well as Pod cgroup sizing.

Consider running the given example workload, test-pod:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  runtimeClassName: kata-fc
  containers:
  - name: busybox-ctr
    image: busybox
    stdin: true
    tty: true
    resources:
      limits:
        cpu: 500m
        memory: 100Mi
  - name: nginx-ctr
    image: nginx
    resources:
      limits:
        cpu: 1500m
        memory: 100Mi

> ```

Reviewer 2: This _might_ work better as a new Task page, “Define a RuntimeClass with a PodOverhead”

At admission time the RuntimeClass [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/)

> will update the workload's PodSpec to include the Overhead as described in the RuntimeClass. If the PodSpec already has this field defined,

Reviewer 2: nit:
```suggestion
updates the workload's PodSpec to include the `overhead` as described in the RuntimeClass. If the PodSpec already has this field defined,
```

> the Pod will be rejected. In the given example, since only the RuntimeClass name is specified, the Pod will be mutated to include the Overhead

Reviewer 2: nit:
```suggestion
the Pod will be rejected. In the given example, since only the RuntimeClass name is specified,
the admission controller mutates the Pod to include an `overhead`
```

field within the PodSpec.

After the RuntimeClass admission controller, the PodSpec will show:

```bash
$ kubectl get pod test-pod -o json | jq '.spec.overhead'
{
  "cpu": "250m",
  "memory": "120Mi"
}

> ```

Reviewer 2: [Separate commands from their output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)

If a ResourceQuota is defined, the sum of container requests as well as the

> Overhead field are counted.

Reviewer 2: nit:
```suggestion
`overhead` field are counted.
```

When kube-scheduler runs, the Overhead as well as sum of container requests are taken into account when identifying

> an appropriate node binding, again seeking a node which has 2.25 CPU and 320 MB of memory available.

Reviewer 2: nit; the units are MiB
```suggestion
an appropriate node binding, again seeking a node which has 2.25 CPU and 320 MiB of memory available.
```

Once scheduled for the node, as part of the Pod creation process Kubelet will create a Pod cgroup (for non burstable QoS)

> and this will be sized as the sum of container (requests or limits) plus the Overhead defined in the PodSpec. 

Reviewer 2: Is this correct?
```suggestion
Once a Pod is scheduled to a node, and that Pod has non-burstable QoS, then the kubelet on that node creates
a new {{< glossary_tooltip text="cgroup" term_id="cgroup" >}} for the Pod. The kubelet instructs the container
runtime to create containers for the Pod inside that cgroup. The cgroup is sized as the sum of container
resources (requests or limits, for resources CPU and memory), plus the `overhead` defined in the PodSpec.
```

If not, please consider rewording to make it clear what happens. This detail is important.

> Taking a look at the container requests, we see a total of 2000m CPU and 200Mi:

Reviewer 2: [Avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we)

```bash
$ kubectl get pod test-pod -o json | jq '.spec.containers[].resources.limits'
{
  "cpu": "500m",
  "memory": "100Mi"
}
{
  "cpu": "1500m",
  "memory": "100Mi"
}
```

> From a node perspective, it sees that 2250m CPU and 320Mi are being requested:

Reviewer 2: nit:
```suggestion
From a node perspective, it sees that 2250m CPU and 320MiB of memory are being requested:
```

```bash
$ kubectl describe node | grep test-pod -B2
  Namespace                   Name                CPU Requests  CPU Limits   Memory Requests  Memory Limits  AGE
  ---------                   ----                ------------  ----------   ---------------  -------------  ---
  default                     test-pod            2250m (56%)   2250m (56%)  320Mi (1%)       320Mi (1%)     36m
```

This is reflected in the cgroups on the node where the owrkload is running. Kubelet creates the Pod cgroup

> with the Pod Overhead taken into account, and it is within this cgroup that the containers run, whether it be through

Reviewer 2: nit:
```suggestion
This is reflected in the cgroups on the node where the workload is running. The kubelet creates the Pod cgroup
with the PodOverhead taken into account, and it is within this cgroup that the containers run, whether it be through
```

containers running inside of a virtual machine or via namespaces containers running on the host.

The relevant memory cgroups for the example workload are shown below:
```bash
root@kind-worker2:~# cat /sys/fs/cgroup/memory/kubepods/pod083eaf28-d8ff-4e1b-b9eb-46d45fffee95/memory.limit_in_bytes
335544320

root@kind-worker2:~# cat /sys/fs/cgroup/memory/kubepods/pod083eaf28-d8ff-4e1b-b9eb-46d45fffee95/**/memory.limit_in_bytes
104857600
9223372036854771712
104857600
```

Ignoring the pause container, which doesn't have a limit set (9223372036854771712), the containers are each 100 MB
and the Pod cgroup is set to 320 MB, as expected.

### Observability

> A `kube_pod_overhead` metric is available in [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) to help identify when `PodOverhead` is

Reviewer 2: nit:
```suggestion
A `kube_pod_overhead` metric is available in [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) to help identify when PodOverhead is
```

being utilized and to help observe stability of workloads running with a defined Overhead.
This functionality is not available in the 1.9 release of kube-state-metrics, but is expected

> in the next release. Users will need to build kube-state-metrics from source in the meantime.

Reviewer 2: [Avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future)

Author: In this case, I'm highlighting that post 1.9 this should be expected.  I could just say not available in 1.9, but I think setting expectation that 2.0 or 1.10 should include *may* be helpful?

        cpu: "250m"
```

Workloads which are created which specify the `kata-fc` RuntimeClass handler will take the memory and
cpu overheads into account for resource quota calculations, node scheduling, as well as Pod cgroup sizing.

Consider running the given example workload, test-pod:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  runtimeClassName: kata-fc
  containers:
  - name: busybox-ctr
    image: busybox
    stdin: true
    tty: true
    resources:
      limits:
        cpu: 500m
        memory: 100Mi
  - name: nginx-ctr
    image: nginx
    resources:
      limits:
        cpu: 1500m
        memory: 100Mi
```

At admission time the RuntimeClass [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/)
updates the workload's PodSpec to include the `overhead` as described in the RuntimeClass. If the PodSpec already has this field defined,
the Pod will be rejected. In the given example, since only the RuntimeClass name is specified, the admission controller mutates the Pod
to include an `overhead`.

After the RuntimeClass admission controller, you can check the updated PodSpec:

```bash

> $ kubectl get pod test-pod -o json | jq '.spec.overhead'

Author: jsonpath output is pretty hard on the eyes, but I understand. Will update.

defined in the PodSpec.

For CPU, if the Pod is Quaranteed or Bustrable QoS, kubelet will set cpu.shares based on the sum of container
requests plus the `overhead` defined in the PodSpec.

Looking at our example, verify the container requests for the workload:
```bash
$ kubectl get pod test-pod -o json | jq '.spec.containers[].resources.limits'
```

The total container requests are 2000m CPU and 200MiB of memory:
```bash
{
  "cpu": "500m",
  "memory": "100Mi"
}
{
  "cpu": "1500m",
  "memory": "100Mi"
}
```

Check this against what is observed by the node:
```bash
$ kubectl describe node | grep test-pod -B2
```

The output shows 2250m CPU and 320MiB of memory are requested, which includes PodOverhead:
```bash
  Namespace                   Name                CPU Requests  CPU Limits   Memory Requests  Memory Limits  AGE
  ---------                   ----                ------------  ----------   ---------------  -------------  ---
  default                     test-pod            2250m (56%)   2250m (56%)  320Mi (1%)       320Mi (1%)     36m
```

Check the Pod's memory cgroups on the node where the workload is running.

Verify the pod level cgroup:
```bash

> root@kind-worker2:~# cat /sys/fs/cgroup/memory/kubepods/pod083eaf28-d8ff-4e1b-b9eb-46d45fffee95/memory.limit_in_bytes

Author: Added some details to clarify this.

> ## Pod Overhead

Reviewer 2: ```suggestion

```

In Kubernetes, the Pod's overhead is set at
[admission](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks)
time according to the overhead associated with the Pod's
[RuntimeClass](/docs/concepts/containers/runtime-class/).

When Pod Overhead is enabled, the overhead is considered in addition to the sum of container
resource requests when scheduling a Pod. Similarly, Kubelet will include the Pod overhead when sizing
the Pod cgroup, and when carrying out Pod eviction ranking.

> ### Set Up

Reviewer 2: ```suggestion
## Enabling Pod overhead {#set-up}
```

You need to make sure that the `PodOverhead`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is enabled (it is on by default as of 1.18)
across your cluster, and a `RuntimeClass` is utilized which defines the `Overhead` field.

> ### Usage example

Reviewer 2: ```suggestion
## Usage example
```

```

The total container requests are 2000m CPU and 200MiB of memory:
```
map[cpu: 500m memory:100Mi] map[cpu:1500m memory:100Mi]
```

Check this against what is observed by the node:
```bash
 kubectl describe node | grep test-pod -B2
```

The output shows 2250m CPU and 320MiB of memory are requested, which includes PodOverhead:
```bash
  Namespace                   Name                CPU Requests  CPU Limits   Memory Requests  Memory Limits  AGE
  ---------                   ----                ------------  ----------   ---------------  -------------  ---
  default                     test-pod            2250m (56%)   2250m (56%)  320Mi (1%)       320Mi (1%)     36m
```

Check the Pod's memory cgroups on the node where the workload is running. First, on the particular node, determine
the Pod identifier:

```bash
sudo crictl pods | grep test-pod
```

The value observed is `7ccf55aee35dd`:
```
7ccf55aee35dd       58 minutes ago      Ready               test-pod                              default             0
```

From this, you can determine the cgroup path for the Pod:
```bash
sudo crictl inspectp -o=json 7ccf55aee35dd | grep cgroupsPath
```

> The resulting cgroup path includes the Pod's `pause` container. We are interested in the Pod cgroup level, which is kubepods/pod*.

Reviewer 2: [Avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we) if you can.

> ```bash

Reviewer 2: Remove marking this as bash code; it's sample output.

        "cgroupsPath": "/kubepods/podd7f4b509-cf94-4951-9417-d1087c92a5b2/7ccf55aee35dd16aca4189c952d83487297f3cd760f1bbf09620e206e7d0c27a"
```

Verify the pod level cgroup:
```bash

> root@kind-worker2:~# cat /sys/fs/cgroup/memory/kubepods/podd7f4b509-cf94-4951-9417-d1087c92a5b2/memory.limit_in_bytes

Reviewer 2: This isn't valid bash code.

You could revise this along the lines of:
```suggestion
# Run this inside the node where the Pod is scheduled
# 
# Also, change the name of the cgroup to match the cgroup that your container runtime allocated
cat /sys/fs/cgroup/memory/kubepods/podd7f4b509-cf94-4951-9417-d1087c92a5b2/memory.limit_in_bytes
```

something like that?
